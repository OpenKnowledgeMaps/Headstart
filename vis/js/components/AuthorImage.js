import React from "react";
import { connect } from "react-redux";

import AuthorImageTemplate from "../templates/AuthorImage";

import defaultImage from "../../images/author_default.png";

export const AuthorImage = ({ hidden, imageLink = "" }) => {
  if (hidden) {
    return null;
  }

  let link = defaultImage;
  if (imageLink !== "") {
    link = imageLink;
  }

  return <AuthorImageTemplate link={link} />;
};

const mapStateToProps = (state) => ({
  hidden: !state.contextLine.showAuthor,
  imageLink: state.contextLine.author.imageLink,
});

export default connect(mapStateToProps)(AuthorImage);
