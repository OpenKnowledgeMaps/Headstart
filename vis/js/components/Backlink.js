import React from "react";
import { connect } from "react-redux";
import { zoomOut } from "../actions";
import { STREAMGRAPH_MODE } from "../reducers/chartType";

import BacklinkTemplate from "../templates/Backlink";

import { createAnimationCallback } from "../utils/eventhandlers";

export const Backlink = ({
  hidden = false,
  streamgraph = false,
  onClick,
  localization = {},
}) => {
  if (hidden) {
    return null;
  }

  const handleOnClick = () => {
    if (onClick && typeof onClick === "function") {
      onClick();
    }
  };

  return (
    <BacklinkTemplate
      label={localization.backlink}
      onClick={handleOnClick}
      className={streamgraph ? "backlink backlink-streamgraph" : "backlink"}
    />
  );
};

const mapStateToProps = (state) => ({
  localization: state.localization,
  hidden: !state.zoom,
  streamgraph: state.chartType === STREAMGRAPH_MODE,
});

const mapDispatchToProps = (dispatch) => ({
  onClick: () => dispatch(zoomOut(createAnimationCallback(dispatch))),
});

export default connect(mapStateToProps, mapDispatchToProps)(Backlink);
