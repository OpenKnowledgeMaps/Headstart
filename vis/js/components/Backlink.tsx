
import React from "react";
import { connect } from "react-redux";
import { zoomOut } from "../actions";
import { STREAMGRAPH_MODE } from "../reducers/chartType";

import BackLinkTemplate from "../templates/Backlink";

import { createAnimationCallback } from "../utils/eventhandlers";
import useMatomo from "../utils/useMatomo";
import { Dispatch } from "redux";

export interface BackLinkProps {
  hidden?: boolean;
  streamgraph?: boolean;
  onClick: () => void;
  localization: any
}

export const BackLink = ({
  hidden = false,
  streamgraph = false,
  onClick,
  localization = {},
}: BackLinkProps) => {
  const { trackEvent } = useMatomo();

  if (hidden) {
    return null;
  }

  const handleOnClick = () => {
    trackEvent("Title & Context line", "Zoom out", "Backlink");
    if (onClick && typeof onClick === "function") {
      onClick();
    }
  };

  return (
    <BackLinkTemplate
      label={localization.backlink}
      onClick={handleOnClick}
      className={streamgraph ? "backlink backlink-streamgraph" : "backlink"}
    />
  );
};

const mapStateToProps = (state: any) => ({
  localization: state.localization,
  hidden: !state.zoom,
  streamgraph: state.chartType === STREAMGRAPH_MODE,
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  onClick: () => dispatch(zoomOut(createAnimationCallback(dispatch))),
});

export default connect(mapStateToProps, mapDispatchToProps)(BackLink);
