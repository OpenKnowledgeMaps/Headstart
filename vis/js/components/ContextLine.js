import React from "react";
import { connect } from "react-redux";

import ContextLineTemplate from "../templates/ContextLine";
import Author from "../templates/contextfeatures/Author";
import DocumentTypes from "../templates/contextfeatures/DocumentTypes";
import NumArticles from "../templates/contextfeatures/NumArticles";
import DataSource from "../templates/contextfeatures/DataSource";
import Timespan from "../templates/contextfeatures/Timespan";
import PaperCount from "../templates/contextfeatures/PaperCount";
import DatasetCount from "../templates/contextfeatures/DatasetCount";
import Funder from "../templates/contextfeatures/Funder";
import ProjectRuntime from "../templates/contextfeatures/ProjectRuntime";
import LegacySearchLang from "../templates/contextfeatures/LegacySearchLang";
import SearchLang from "../templates/contextfeatures/SearchLang";
import Timestamp from "../templates/contextfeatures/Timestamp";
import Modifier from "../templates/contextfeatures/Modifier";
import MoreInfoLink from "../templates/contextfeatures/MoreInfoLink";
import MetadataQuality from "../templates/contextfeatures/MetadataQuality";
import ContextTimeFrame from "../templates/contextfeatures/ContextTimeFrame";
import DocumentLang from "../templates/contextfeatures/DocumentLang";

const defined = (param) => param !== undefined && param !== null;

/**
 * Component that creates the heading contextline.
 *
 * It has to be a class component because of the popovers (they use 'this').
 */
class ContextLine extends React.Component {
  render() {
    const { params, localization, hidden, service } = this.props;
    const { popoverContainer } = this.props;

    if (hidden) {
      return null;
    }

    return (
      <ContextLineTemplate>
          {params.showAuthor && (
              <Author
                  bioLabel={localization.bio_link}
                  livingDates={params.author.livingDates}
                  link={"https://d-nb.info/gnd/" + params.author.id}
              />
          )}
          <NumArticles
              articlesCount={params.articlesCount}
              openAccessArticlesCount={params.openAccessCount}
              articlesCountLabel={localization.articles_label}
          >
              <Modifier popoverContainer={popoverContainer}/>
          </NumArticles>
          {defined(params.dataSource) && (
              <DataSource
                  label={localization.source_label}
                  source={params.dataSource}
                  contentProvider={params.contentProvider}
                  popoverContainer={this.props.popoverContainer}
              />
          )}
          {defined(params.timespan) &&
              <Timespan>
                  <ContextTimeFrame popoverContainer={popoverContainer} timespan={params.timespan}/>
              </Timespan>}
          <DocumentTypes
              documentTypes={params.documentTypes}
              popoverContainer={popoverContainer}
          />
          {/* was an issue to left "All Languages" as default value in the context if no lang_id in parameters */}
          <DocumentLang
              value={params.documentLang}
              popoverContainer={popoverContainer}
          />
          {defined(params.paperCount) && (
              <PaperCount
                  value={params.paperCount}
                  label={localization.paper_count_label}
              />
          )}
          {defined(params.datasetCount) && (
              <DatasetCount
                  value={params.datasetCount}
                  label={localization.dataset_count_label}
              />
          )}
          {defined(params.funder) && <Funder>{params.funder}</Funder>}
          {defined(params.projectRuntime) && (
              <ProjectRuntime>{params.projectRuntime}</ProjectRuntime>
          )}
          {defined(params.legacySearchLanguage) && (
              <LegacySearchLang>{params.legacySearchLanguage}</LegacySearchLang>
          )}
          {defined(params.timestamp) && (
              <Timestamp
                  value={params.timestamp}
                  label={localization.timestamp_label}
              />
          )}
          <MetadataQuality
              quality={params.metadataQuality}
              service={service}
              popoverContainer={popoverContainer}
          />
          {defined(params.searchLanguage) && (
              <SearchLang>{params.searchLanguage}</SearchLang>
          )}
          <MoreInfoLink/>
      </ContextLineTemplate>
    );
  }
}

const mapStateToProps = (state) => ({
  hidden: state.zoom || !state.contextLine.show,
  params: {
    ...state.contextLine,
    timespan: state.timespan,
  },
  service: state.service,
  localization: state.localization,
});

export default connect(mapStateToProps)(ContextLine);
