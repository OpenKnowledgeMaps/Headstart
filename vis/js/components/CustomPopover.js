import React from "react";

import { OverlayTrigger, Popover } from "react-bootstrap";

const CustomPopover = ({
  id,
  content,
  children,
  container, // = null
  placement = "bottom",
}) => {
  const popover = <Popover id={id}>{content}</Popover>;

  return (
    <OverlayTrigger
      trigger={["hover", "focus"]}
      placement={placement}
      overlay={popover}
      container={container}
    >
      {children}
    </OverlayTrigger>
  );
};

export default CustomPopover;
