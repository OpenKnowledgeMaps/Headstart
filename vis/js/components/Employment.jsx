import React, { useEffect, useRef, useState } from "react";
import HoverPopover from "../templates/HoverPopover";

export function Employment({ author, popoverContainer }) {
  const authorRoleId = "author-role";
  const authorOrganizationId = "author-organization";

  const authorRoleTextRef = useRef();
  const [authorRoleHasOverflow, setAuthorRoleHasOverflow] = useState(false);
  const authorOrganizationTextRef = useRef();
  const [authorOrganizationHasOverflow, setAuthorOrganizationHasOverflow] =
    useState(false);

  useEffect(() => {
    const checkOverflow = () => {
      const authorRoleText = authorRoleTextRef.current;
      
      if (!authorRoleText) {
        return;
      }
      
      if (authorRoleText && authorRoleText.scrollWidth > authorRoleText.clientWidth) {
        setAuthorRoleHasOverflow(true);
      } else {
        setAuthorRoleHasOverflow(false);
      }
      
      const authorOrganizationText = authorOrganizationTextRef.current;

      if (
        authorOrganizationText && authorOrganizationText.scrollWidth > authorOrganizationText.clientWidth
      ) {
        setAuthorOrganizationHasOverflow(true);
      } else {
        setAuthorOrganizationHasOverflow(false);
      }
    };

    // Check for overflow on load and on window resize
    checkOverflow();
    window.addEventListener("resize", checkOverflow);

    // Clean up event listener on component unmount
    return () => {
      window.removeEventListener("resize", checkOverflow);
    };
  }, []);

  return (
    <>
      {author?.employment?.role ? (
        <span id={authorRoleId} className="context_item shrinkable">
          {authorRoleHasOverflow ? (
            <HoverPopover
              id={`${authorRoleId}-popover`}
              container={popoverContainer}
              content={author.employment.role}
            >
              <span
                ref={authorRoleTextRef}
                className={`context_moreinfo overflow-ellipsis ${authorRoleHasOverflow ? 'has-overflow' : ''}`}
              >
                {author.employment.role}
              </span>
            </HoverPopover>
          ) : (
            <span
              ref={authorRoleTextRef}
              className="context_moreinfo overflow-ellipsis"
            >
              {author.employment.role}
            </span>
          )}
        </span>
      ) : null}

      {author?.employment?.organization ? (
        <span id={authorOrganizationId} className="context_item shrinkable">
          {authorOrganizationHasOverflow ? (
            <HoverPopover
              id={`${authorOrganizationId}-popover`}
              container={popoverContainer}
              content={author.employment.organization}
            >
              <span
                ref={authorOrganizationTextRef}
                className={`context_moreinfo overflow-ellipsis ${authorOrganizationHasOverflow ? 'has-overflow' : ''}`}
              >
                {author.employment.organization}
              </span>
            </HoverPopover>
          ) : (
            <span
              ref={authorOrganizationTextRef}
              className="context_moreinfo overflow-ellipsis"
            >
              {author.employment.organization}
            </span>
          )}
        </span>
      ) : null}
    </>
  );
}
