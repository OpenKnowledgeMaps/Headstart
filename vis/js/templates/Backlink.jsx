import React from "react";

const Backlink = ({ label, onClick }) => {
  return (
    // html template starts here
    <p id="backlink" className="backlink">
      <a className="underline" onClick={onClick}>
        {label}
      </a>
    </p>
    // html template ends here
  );
};

export default Backlink;
