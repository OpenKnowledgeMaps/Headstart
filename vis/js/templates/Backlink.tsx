import React from "react";

export interface BackLinkProps {
  label: string;
  onClick: () => void;
  className?: string;
}

const BackLink = ({ label, onClick, className = "backlink" }: BackLinkProps) => {
  return (
    <p id="backlink" className={className}>
      <a className="underline" onClick={onClick}>
        {label}
      </a>
    </p>
  );
};

export default BackLink;
