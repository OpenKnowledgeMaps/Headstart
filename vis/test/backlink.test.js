import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { shallow } from "enzyme";

import { Backlink } from "../js/components/Backlink";

const setup = (overrideProps) => {
  const props = Object.assign(
    {
      hidden: false,
      onClick: jest.fn(),
    },
    overrideProps
  );

  const wrapper = shallow(<Backlink {...props} />);

  return { wrapper, props };
};

/**
 * Redux state tests are missing.
 * They probably belong in a separate file...
 */

/**
 * All the tests for the Backlink component, including its template, can be found here.
 */
describe("Backlink component", () => {
  /**
   * Basic render test
   *   using Enzyme for shallow rendering (no children rendered)
   */
  it("renders", () => {
    const { wrapper } = setup();
    expect(wrapper.exists()).toBe(true);
  });

  /**
   * Component render tests
   *   using Enzyme again, so just shallow rendering
   */
  describe("visibility", () => {
    it("renders shown", () => {
      const { wrapper } = setup({ hidden: false });
      expect(wrapper.exists()).toBe(true);
    });

    it("renders hidden", () => {
      const { wrapper } = setup({ hidden: true });
      expect(wrapper.type()).toEqual(null);
    });
  });

  /**
   * Component event tests
   *   using react-dom, because we need full DOM for that
   *
   * More component integration tests can be added here too.
   */
  describe("onclick", () => {
    // suppressing the info warns in the backlink onClick handler
    global.console = {
      log: jest.fn(),
      warn: jest.fn(),
      error: console.error,
      info: console.info,
      debug: console.debug,
    };

    let container = null;
    beforeEach(() => {
      container = document.createElement("div");
      document.body.appendChild(container);
    });

    afterEach(() => {
      unmountComponentAtNode(container);
      container.remove();
      container = null;
    });

    it("triggers the onClick function when clicked", () => {
      const onClick = jest.fn();
      act(() => {
        render(<Backlink onClick={onClick} />, container);
      });

      const link = document.querySelector("a.underline");
      act(() => {
        link.dispatchEvent(new MouseEvent("click", { bubbles: true }));
      });

      expect(onClick).toHaveBeenCalledTimes(1);
    });

    it("doesn't crash when onClick is not a function", () => {
      const onClick = null;
      act(() => {
        render(<Backlink onClick={onClick} />, container);
      });

      const link = document.querySelector("a.underline");
      act(() => {
        link.dispatchEvent(new MouseEvent("click", { bubbles: true }));
      });
    });
  });

  /**
   * Redux/React integration tests probably belong here...
   */
});
