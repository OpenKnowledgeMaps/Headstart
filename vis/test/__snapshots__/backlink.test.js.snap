// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backlink component renders as snapshot 1`] = `
<p
  className="backlink"
  id="backlink"
>
  <a
    className="underline"
    onClick={[Function]}
  >
    ‚Üê Back to overview
  </a>
</p>
`;
